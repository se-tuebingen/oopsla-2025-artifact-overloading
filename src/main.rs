use overloading::compile_expr;
use overloading::syntax::parse_expr;

use clap::Parser;

#[derive(Parser)]
#[command(name = "overloading")]
struct Args {
    /// Input file to process
    filename: String,

    /// Use short output format
    #[arg(short, long)]
    short: bool,
}

/// Main entry point for the native CLI.
fn main() {
    let args = Args::parse();

    let content = std::fs::read_to_string(&args.filename).expect("Failed to read file");
    match parse_expr(&content) {
        Ok(expr) => {
            println!("{}", compile_expr(&expr, args.short));
        }
        Err(err) => {
            panic!("Failed to parse expression: {}", err);
        }
    }
}

/// Snapshot tests for the compiler examples.
#[cfg(test)]
#[test]
fn test_compiler_examples() {
    use insta::{assert_json_snapshot, glob, with_settings};

    glob!("../", "examples/*.over", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let output = match parse_expr(&input) {
            Ok(expr) => overloading::compile_expr_json(&expr, true),
            Err(err) => overloading::json::JsonProcessingResult::InferenceError {
                expression: input,
                error: err,
            },
        };

        // Use the filename (without extension) as snapshot suffix
        let filename = path.file_stem().unwrap().to_str().unwrap();
        with_settings!({
            snapshot_path => "../examples/snapshots",
            snapshot_suffix => filename,
            sort_maps => true,
        }, {
            assert_json_snapshot!(output);
        });
    });
}
