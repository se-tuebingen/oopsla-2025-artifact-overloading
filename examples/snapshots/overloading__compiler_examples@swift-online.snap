---
source: src/main.rs
expression: output
input_file: examples/swift-online.over
---
{
  "status": "success",
  "expression": "assume addInt: (Int, Int) => Int\nassume addString: (String, String) => String\nassume addDouble: (Double, Double) => Double\nlet address = \"127.0.0.1\"\nlet username = \"steve\"\nlet password = \"1234\"\nlet channel = 11\na‹addString, addInt, addDouble›(\n  \"http://\",\n  b‹addString, addInt, addDouble›(\n    username,\n    c‹addString, addInt, addDouble›(\n      \":\",\n      d‹addString, addInt, addDouble›(\n        password,\n        e‹addString, addInt, addDouble›(\n          \"@\",\n          f‹addString, addInt, addDouble›(\n            address,\n            g‹addString, addInt, addDouble›(\n              \"/api/\",\n              h‹addString, addInt, addDouble›(channel, \"/picture\")\n            )\n          )\n        )\n      )\n    )\n  )\n)",
  "dimension_spans": "[spans]",
  "inferred_type": "app_15",
  "worlds": {
    "type": "no"
  },
  "constraints": [
    {
      "from": "(String, String) => String",
      "choices": [
        "a₁"
      ],
      "to": "over_a_0"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "a₂"
      ],
      "to": "over_a_0"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "a₃"
      ],
      "to": "over_a_0"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "b₁"
      ],
      "to": "over_b_1"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "b₂"
      ],
      "to": "over_b_1"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "b₃"
      ],
      "to": "over_b_1"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "c₁"
      ],
      "to": "over_c_2"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "c₂"
      ],
      "to": "over_c_2"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "c₃"
      ],
      "to": "over_c_2"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "d₁"
      ],
      "to": "over_d_3"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "d₂"
      ],
      "to": "over_d_3"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "d₃"
      ],
      "to": "over_d_3"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "e₁"
      ],
      "to": "over_e_4"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "e₂"
      ],
      "to": "over_e_4"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "e₃"
      ],
      "to": "over_e_4"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "f₁"
      ],
      "to": "over_f_5"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "f₂"
      ],
      "to": "over_f_5"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "f₃"
      ],
      "to": "over_f_5"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "g₁"
      ],
      "to": "over_g_6"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "g₂"
      ],
      "to": "over_g_6"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "g₃"
      ],
      "to": "over_g_6"
    },
    {
      "from": "(String, String) => String",
      "choices": [
        "h₁"
      ],
      "to": "over_h_7"
    },
    {
      "from": "(Int, Int) => Int",
      "choices": [
        "h₂"
      ],
      "to": "over_h_7"
    },
    {
      "from": "(Double, Double) => Double",
      "choices": [
        "h₃"
      ],
      "to": "over_h_7"
    },
    {
      "from": "over_h_7",
      "choices": [],
      "to": "(Int, String) => app_8"
    },
    {
      "from": "over_g_6",
      "choices": [],
      "to": "(String, app_8) => app_9"
    },
    {
      "from": "over_f_5",
      "choices": [],
      "to": "(String, app_9) => app_10"
    },
    {
      "from": "over_e_4",
      "choices": [],
      "to": "(String, app_10) => app_11"
    },
    {
      "from": "over_d_3",
      "choices": [],
      "to": "(String, app_11) => app_12"
    },
    {
      "from": "over_c_2",
      "choices": [],
      "to": "(String, app_12) => app_13"
    },
    {
      "from": "over_b_1",
      "choices": [],
      "to": "(String, app_13) => app_14"
    },
    {
      "from": "over_a_0",
      "choices": [],
      "to": "(String, app_14) => app_15"
    }
  ],
  "bounds": [
    {
      "variable": "app_10",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "f₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "f₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "f₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "e₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "e₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "e₁"
          ]
        }
      ]
    },
    {
      "variable": "app_11",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "e₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "e₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "e₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "d₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "d₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "d₁"
          ]
        }
      ]
    },
    {
      "variable": "app_12",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "d₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "d₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "d₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "c₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "c₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "c₁"
          ]
        }
      ]
    },
    {
      "variable": "app_13",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "c₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "c₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "c₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "b₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "b₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "b₁"
          ]
        }
      ]
    },
    {
      "variable": "app_14",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "b₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "b₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "b₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "a₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₁"
          ]
        }
      ]
    },
    {
      "variable": "app_15",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "a₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": []
    },
    {
      "variable": "app_8",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "h₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "h₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "h₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "g₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "g₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "g₁"
          ]
        }
      ]
    },
    {
      "variable": "app_9",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "g₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "g₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "g₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "f₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "f₂"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "f₁"
          ]
        }
      ]
    },
    {
      "variable": "over_a_0",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "a₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "a₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(String, app_14) => app_15",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_b_1",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "b₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "b₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "b₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(String, app_13) => app_14",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_c_2",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "c₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "c₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "c₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(String, app_12) => app_13",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_d_3",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "d₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "d₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "d₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(String, app_11) => app_12",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_e_4",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "e₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "e₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "e₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(String, app_10) => app_11",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_f_5",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "f₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "f₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "f₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(String, app_9) => app_10",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_g_6",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "g₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "g₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "g₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(String, app_8) => app_9",
          "choices": []
        }
      ]
    },
    {
      "variable": "over_h_7",
      "lower": [
        {
          "type_name": "(Double, Double) => Double",
          "choices": [
            "h₃"
          ]
        },
        {
          "type_name": "(Int, Int) => Int",
          "choices": [
            "h₂"
          ]
        },
        {
          "type_name": "(String, String) => String",
          "choices": [
            "h₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(Int, String) => app_8",
          "choices": []
        }
      ]
    }
  ],
  "errors": [
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "a₁",
          "b₃"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "a₂",
          "b₃"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "b₁",
          "c₃"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "b₂",
          "c₃"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "c₁",
          "d₃"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "c₂",
          "d₃"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "d₁",
          "e₃"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "d₂",
          "e₃"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "e₁",
          "f₃"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "e₂",
          "f₃"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "f₁",
          "g₃"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "f₂",
          "g₃"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "g₁",
          "h₃"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Double",
        "choices": [
          "g₂",
          "h₃"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "a₁",
          "b₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "a₃",
          "b₂"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "b₁",
          "c₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "b₃",
          "c₂"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "c₁",
          "d₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "c₃",
          "d₂"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "d₁",
          "e₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "d₃",
          "e₂"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "e₁",
          "f₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "e₃",
          "f₂"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "f₁",
          "g₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "f₃",
          "g₂"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "g₁",
          "h₂"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "g₃",
          "h₂"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "h₁"
        ],
        "to": "String"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "h₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "a₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "a₂",
          "b₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "a₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "a₃",
          "b₁"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "b₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "b₂",
          "c₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "b₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "b₃",
          "c₁"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "c₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "c₂",
          "d₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "c₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "c₃",
          "d₁"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "d₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "d₂",
          "e₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "d₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "d₃",
          "e₁"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "e₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "e₂",
          "f₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "e₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "e₃",
          "f₁"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "f₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "f₂",
          "g₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "f₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "f₃",
          "g₁"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "g₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "g₂",
          "h₁"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "g₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "g₃",
          "h₁"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "h₂"
        ],
        "to": "Int"
      },
      "message": "Incompatible base types"
    },
    {
      "constraint": {
        "from": "String",
        "choices": [
          "h₃"
        ],
        "to": "Double"
      },
      "message": "Incompatible base types"
    }
  ]
}
